conflict between prefer-const and no-use-before-define (2.5.0 regression)

**What version of ESLint are you using?**
2.5.0 (I think my previous version that didn't fail was 2.4.0, but it could have been a different 2.x)
**What parser (default, Babel-ESLint, etc.) are you using?**
Default
**Please show your full configuration:**
```js
{
...
"no-use-before-define": ["error", "nofunc"],
...
 "prefer-const": "error",
}
```
**What did you do? Please include the actual source code causing the issue.**
```js
            let interval;
            function checkForWindow() {
                const newFocussed = BrowserWindow.getFocusedWindow();
                if (newFocussed !== currentlyFocussed) {
                    clearInterval(interval);
                    newFocussed.toggleDevTools();
                }
            }
            interval = setInterval(checkForWindow, msIntervalToCheckForWindow);
```
**What did you expect to happen?**
no warnings
**What actually happened? Please include the actual, raw output from ESLint.**
either
```sh
  76:13  error  'interval' is never reassigned, use 'const' instead  prefer-const
```
or if I change it to const after the interval
```sh
  72:35  error  'interval' was used before it was defined    no-use-before-define
```

I see the rule says this is an error
```js
// the initializer is separated.
let a;
a = 0;
console.log(a);
```
but this is not
```js
let a;
if (true) {
    a = 0;
}
console.log(a);
```
I would suggest that
```js
let a;
// read or write of a in any sub block (e.g. if or function)
console.log(a);
```
is not an error?

--- a/docs/rules/prefer-const.md
+++ b/docs/rules/prefer-const.md
@@ -88,7 +88,10 @@ console.log(b);
 
 ```json
 {
-    "prefer-const": ["error", {"destructuring": "any"}]
+    "prefer-const": ["error", {
+        "destructuring": "any",
+        "ignoreReadBeforeAssign": false
+    }]
 }
 ```
 
@@ -148,6 +151,41 @@ let {a, b} = obj;
 a = a + 1;
 ```
 
+### ignoreReadBeforeAssign
+
+This is an option to avoid conflicting with `no-use-before-define` rule (without `"nofunc"` option).
+If `true` is specified, this rule will ignore variables that are read between the declaration and the first assignment.
+Default is `false`.
+
+Examples of **correct** code for the `{"ignoreReadBeforeAssign": true}` option:
+
+```js
+/*eslint prefer-const: ["error", {"ignoreReadBeforeAssign": true}]*/
+/*eslint-env es6*/
+
+let timer;
+function initialize() {
+    if (foo()) {
+        clearInterval(timer);
+    }
+}
+timer = setInterval(initialize, 100);
+```
+
+Examples of **correct** code for the defaut `{"ignoreReadBeforeAssign": false}` option:
+
+```js
+/*eslint prefer-const: ["error", {"ignoreReadBeforeAssign": false}]*/
+/*eslint-env es6*/
+
+const timer = setInterval(initialize, 100);
+function initialize() {
+    if (foo()) {
+        clearInterval(timer);
+    }
+}
+```
+
 ## When Not To Use It
 
 If you don't want to be notified about variables that are never reassigned after initial assignment, you can safely disable this rule.
@@ -155,3 +193,4 @@ If you don't want to be notified about variables that are never reassigned after
 ## Related Rules
 
 * [no-var](no-var.md)
+* [no-use-before-define](no-use-before-define.md)

--- a/lib/rules/prefer-const.js
+++ b/lib/rules/prefer-const.js
@@ -62,45 +62,72 @@ function canBecomeVariableDeclaration(identifier) {
 }
 
 /**
- * Gets the WriteReference of a given variable if the variable should be
- * declared as const.
+ * Gets an identifier node of a given variable.
+ *
+ * If the initialization exists or one or more reading references exist before
+ * the first assignment, the identifier node is the node of the declaration.
+ * Otherwise, the identifier node is the node of the first assignment.
+ *
+ * If the variable should not change to const, this function returns null.
+ * - If the variable is reassigned.
+ * - If the variable is never initialized and assigned.
+ * - If the variable is initialized in a different scope from the declaration.
+ * - If the unique assignment of the variable cannot change to a declaration.
  *
  * @param {escope.Variable} variable - A variable to get.
- * @returns {escope.Reference|null} The singular WriteReference or null.
+ * @param {boolean} ignoreReadBeforeAssign -
+ *      The value of `ignoreReadBeforeAssign` option.
+ * @returns {ASTNode|null}
+ *      An Identifier node if the variable should change to const.
+ *      Otherwise, null.
  */
-function getWriteReferenceIfShouldBeConst(variable) {
+function getIdentifierIfShouldBeConst(variable, ignoreReadBeforeAssign) {
     if (variable.eslintUsed) {
         return null;
     }
 
-    // Finds the singular WriteReference.
-    var retv = null;
+    // Finds the unique WriteReference.
+    var writer = null;
+    var isReadBeforeInit = false;
     var references = variable.references;
 
     for (var i = 0; i < references.length; ++i) {
         var reference = references[i];
 
         if (reference.isWrite()) {
-            var isReassigned = Boolean(
-                retv && retv.identifier !== reference.identifier
+            var isReassigned = (
+                writer !== null &&
+                writer.identifier !== reference.identifier
             );
 
             if (isReassigned) {
                 return null;
             }
-            retv = reference;
+            writer = reference;
+
+        } else if (reference.isRead() && writer === null) {
+            if (ignoreReadBeforeAssign) {
+                return null;
+            }
+            isReadBeforeInit = true;
         }
     }
 
-    // Checks the writer is located in the same scope and can be modified to
-    // const.
-    var isSameScopeAndCanBecomeVariableDeclaration = Boolean(
-        retv &&
-        retv.from === variable.scope &&
-        canBecomeVariableDeclaration(retv.identifier)
+    // If the assignment is from a different scope, ignore it.
+    // If the assignment cannot change to a declaration, ignore it.
+    var shouldBeConst = (
+        writer !== null &&
+        writer.from === variable.scope &&
+        canBecomeVariableDeclaration(writer.identifier)
     );
 
-    return isSameScopeAndCanBecomeVariableDeclaration ? retv : null;
+    if (!shouldBeConst) {
+        return null;
+    }
+    if (isReadBeforeInit) {
+        return variable.defs[0].name;
+    }
+    return writer.identifier;
 }
 
 /**
@@ -136,15 +163,17 @@ function getDestructuringHost(reference) {
  * destructuring.
  *
  * @param {escope.Variable[]} variables - Variables to group by destructuring.
- * @returns {Map<ASTNode, (escope.Reference|null)[]>} Grouped references.
+ * @param {boolean} ignoreReadBeforeAssign -
+ *      The value of `ignoreReadBeforeAssign` option.
+ * @returns {Map<ASTNode, ASTNode[]>} Grouped identifier nodes.
  */
-function groupByDestructuring(variables) {
-    var writersMap = new Map();
+function groupByDestructuring(variables, ignoreReadBeforeAssign) {
+    var identifierMap = new Map();
 
     for (var i = 0; i < variables.length; ++i) {
         var variable = variables[i];
         var references = variable.references;
-        var writer = getWriteReferenceIfShouldBeConst(variable);
+        var identifier = getIdentifierIfShouldBeConst(variable, ignoreReadBeforeAssign);
         var prevId = null;
 
         for (var j = 0; j < references.length; ++j) {
@@ -158,20 +187,20 @@ function groupByDestructuring(variables) {
             }
             prevId = id;
 
-            // Add the writer into the destructuring group.
+            // Add the identifier node into the destructuring group.
             var group = getDestructuringHost(reference);
 
             if (group) {
-                if (writersMap.has(group)) {
-                    writersMap.get(group).push(writer);
+                if (identifierMap.has(group)) {
+                    identifierMap.get(group).push(identifier);
                 } else {
-                    writersMap.set(group, [writer]);
+                    identifierMap.set(group, [identifier]);
                 }
             }
         }
     }
 
-    return writersMap;
+    return identifierMap;
 }
 
 //------------------------------------------------------------------------------
@@ -190,7 +219,8 @@ module.exports = {
             {
                 type: "object",
                 properties: {
-                    destructuring: {enum: ["any", "all"]}
+                    destructuring: {enum: ["any", "all"]},
+                    ignoreReadBeforeAssign: {type: "boolean"}
                 },
                 additionalProperties: false
             }
@@ -200,21 +230,20 @@ module.exports = {
     create: function(context) {
         var options = context.options[0] || {};
         var checkingMixedDestructuring = options.destructuring !== "all";
+        var ignoreReadBeforeAssign = options.ignoreReadBeforeAssign === true;
         var variables = null;
 
         /**
-         * Reports a given reference.
+         * Reports a given Identifier node.
          *
-         * @param {escope.Reference} reference - A reference to report.
+         * @param {ASTNode} node - An Identifier node to report.
          * @returns {void}
          */
-        function report(reference) {
-            var id = reference.identifier;
-
+        function report(node) {
             context.report({
-                node: id,
+                node: node,
                 message: "'{{name}}' is never reassigned, use 'const' instead.",
-                data: id
+                data: node
             });
         }
 
@@ -225,30 +254,30 @@ module.exports = {
          * @returns {void}
          */
         function checkVariable(variable) {
-            var writer = getWriteReferenceIfShouldBeConst(variable);
+            var node = getIdentifierIfShouldBeConst(variable, ignoreReadBeforeAssign);
 
-            if (writer) {
-                report(writer);
+            if (node) {
+                report(node);
             }
         }
 
         /**
-         * Reports given references if all of the reference should be declared as
-         * const.
+         * Reports given identifier nodes if all of the nodes should be declared
+         * as const.
          *
-         * The argument 'writers' is an array of references.
-         * This reference is the result of
-         * 'getWriteReferenceIfShouldBeConst(variable)', so it's nullable.
-         * In simple declaration or assignment cases, the length of the array is 1.
-         * In destructuring cases, the length of the array can be 2 or more.
+         * The argument 'nodes' is an array of Identifier nodes.
+         * This node is the result of 'getIdentifierIfShouldBeConst()', so it's
+         * nullable. In simple declaration or assignment cases, the length of
+         * the array is 1. In destructuring cases, the length of the array can
+         * be 2 or more.
          *
-         * @param {(escope.Reference|null)[]} writers - References which are grouped
-         *      by destructuring to report.
+         * @param {(escope.Reference|null)[]} nodes -
+         *      References which are grouped by destructuring to report.
          * @returns {void}
          */
-        function checkGroup(writers) {
-            if (writers.every(Boolean)) {
-                writers.forEach(report);
+        function checkGroup(nodes) {
+            if (nodes.every(Boolean)) {
+                nodes.forEach(report);
             }
         }
 
@@ -261,7 +290,8 @@ module.exports = {
                 if (checkingMixedDestructuring) {
                     variables.forEach(checkVariable);
                 } else {
-                    groupByDestructuring(variables).forEach(checkGroup);
+                    groupByDestructuring(variables, ignoreReadBeforeAssign)
+                        .forEach(checkGroup);
                 }
 
                 variables = null;

--- a/tests/lib/rules/prefer-const.js
+++ b/tests/lib/rules/prefer-const.js
@@ -82,6 +82,13 @@ ruleTester.run("prefer-const", rule, {
             code: "let a, b; ({a, b} = obj); b++;",
             options: [{destructuring: "all"}],
             parserOptions: {ecmaVersion: 6}
+        },
+
+        // ignoreReadBeforeAssign
+        {
+            code: "let x; function foo() { bar(x); } x = 0;",
+            options: [{ignoreReadBeforeAssign: true}],
+            parserOptions: {ecmaVersion: 6}
         }
     ],
     invalid: [
@@ -157,17 +164,17 @@ ruleTester.run("prefer-const", rule, {
         {
             code: "let x; x = 0;",
             parserOptions: { ecmaVersion: 6 },
-            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
+            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier", column: 8}]
         },
         {
             code: "switch (a) { case 0: let x; x = 0; }",
             parserOptions: { ecmaVersion: 6 },
-            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
+            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier", column: 29}]
         },
         {
             code: "(function() { let x; x = 1; })();",
             parserOptions: { ecmaVersion: 6 },
-            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
+            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier", column: 22}]
         },
 
         {
@@ -214,6 +221,13 @@ ruleTester.run("prefer-const", rule, {
             options: [{destructuring: "all"}],
             parserOptions: {ecmaVersion: 6},
             errors: [{ message: "'c' is never reassigned, use 'const' instead.", type: "Identifier"}]
+        },
+
+        // Warnings are located at declaration if there are reading references before assignments.
+        {
+            code: "let x; function foo() { bar(x); } x = 0;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier", column: 5}]
         }
     ]
 });

